src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs

```diff
@@ -87,8 +87,14 @@ private class QueueItem<TRequestType, TResponseType> : IQueueItem
                 RequestTelemetryLogger telemetryLogger,
                 CancellationToken cancellationToken)
             {
-                // Set the tcs state to cancelled if the token gets cancelled outside of our callback (for example the server shutting down).
-                cancellationToken.Register(() => _completionSource.TrySetCanceled(cancellationToken));
+                if (cancellationToken.CanBeCanceled)
+                {
+                    // Set the tcs state to cancelled if the token gets cancelled outside of our callback (for example the server shutting down).
+                    var registration = cancellationToken.Register(() => _completionSource.TrySetCanceled(cancellationToken));
+
+                    // Dispose of the registration as soon as we no longer need it
+                    _completionSource.Task.ContinueWith(_ => registration.Dispose(), CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```

Comment:
Heh, I think you might want to use Unregister here, because it may cause deadlocks: https://source.dot.net/#System.Private.CoreLib/CancellationTokenRegistration.cs,f05a1300415bcfde,references

It's noted very obscurely in the docs [here](https://source.dot.net/#System.Private.CoreLib/CancellationTokenRegistration.cs,f05a1300415bcfde,references):

> If you call [Dispose](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokenregistration.dispose) while the callback is running, and you hold a lock that the callback is waiting on, your program can deadlock. After Dispose returns, you can free any resources required by the callback.

Reply:
I pushed an update to address the straightforward cases in production code.